File: app/src/main/java/at/priv/graf/zazentimer/audio/Audio.java
package at.priv.graf.zazentimer.audio;

import android.content.Context;
import android.media.AudioManager;
import android.media.AudioAttributes;
import android.media.MediaPlayer;
import android.net.Uri;
import android.util.Log;
import at.priv.graf.zazentimer.Bell;
import at.priv.graf.zazentimer.bo.Section;

public class Audio implements MediaPlayer.OnCompletionListener {
    private static final String TAG = "ZMT_Audio";
    private Context context;
    private AudioManager manager;
    private volatile boolean playing;
    private MediaPlayer player = null;

    public Audio(Context context) {
        Log.d(TAG, "New Audio Instance");
        this.context = context;
        this.manager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
    }

    public void release() {
        Log.d(TAG, "Releasing Audio Instance");
        stopAndRelease();
    }

    private MediaPlayer preparePlayer(Bell bell, int volume) {
        Log.d(TAG, "preparing Audio Player");
        Uri uri = bell.getUri();
        if (uri == null) {
            return null;
        }
        MediaPlayer mediaPlayer = new MediaPlayer();
        try {
            mediaPlayer.setAudioAttributes(new AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_ALARM)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build());
            mediaPlayer.setDataSource(this.context, uri);
            mediaPlayer.prepare();
            float f = volume / 100.0f;
            mediaPlayer.setVolume(f, f);
            return mediaPlayer;
        } catch (Exception e) {
            Log.e(TAG, "Error creating MediaPlayer", e);
            return null;
        }
    }

    private void stopAndRelease() {
        Log.d(TAG, "stopAndRelease Audio Instance");
        if (this.player != null) {
            try {
                this.player.stop();
                this.player.reset();
            } catch (Exception e) {
                Log.d(TAG, e.getMessage());
            }
            try {
                this.player.release();
            } catch (Exception e2) {
                Log.d(TAG, e2.getMessage());
            }
            this.player = null;
        }
    }

    public void playAbsVolume(Section section) {
        playAbsVolume(BellCollection.getInstance().getBellForSection(section), section.volume);
    }

    public void playAbsVolume(Bell bell, int volume) {
        if (this.player != null) {
            stopAndRelease();
        }
        this.player = preparePlayer(bell, volume);
        if (this.player != null) {
            this.playing = true;
            Log.d(TAG, "Start playing Bell");
            this.player.setLooping(false);
            this.player.start();
            this.player.setOnCompletionListener(this);
            return;
        }
        Log.e(TAG, "Could not preparePlayer");
    }

    public boolean isPlaying() {
        return this.playing;
    }

    public int getStreamVolume() {
        int streamMaxVolume = this.manager.getStreamMaxVolume(AudioManager.STREAM_ALARM);
        float streamVolume = this.manager.getStreamVolume(AudioManager.STREAM_ALARM);
        int round = Math.round((100.0f * streamVolume) / streamMaxVolume);
        Log.d(TAG, "getStreamVolume: vol=" + streamVolume + " max=" + streamMaxVolume + " = " + round);
        return round;
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        Log.d(TAG, "onCompletion called");
        this.playing = false;
        stopAndRelease();
    }
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/audio/Audio.kt
package at.priv.graf.zazentimer.audio

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Build
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class Audio(
    private val context: Context,
) : MediaPlayer.OnCompletionListener {
    @Volatile
    private var playing: Boolean = false
    private var player: MediaPlayer? = null
    private val manager: AudioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    init {
        Log.d(TAG, "New Audio Instance")
    }

    suspend fun release() {
        Log.d(TAG, "Releasing Audio Instance")
        withContext(Dispatchers.IO) {
            stopAndRelease()
        }
    }

    private fun preparePlayer(
        uri: Uri,
        volume: Int,
    ): MediaPlayer {
        Log.d(TAG, "preparing Audio Player")
        val mediaPlayer = MediaPlayer()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            mediaPlayer.setAudioAttributes(
                AudioAttributes
                    .Builder()
                    .setUsage(AudioAttributes.USAGE_ALARM)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build(),
            )
        } else {
            @Suppress("DEPRECATION")
            mediaPlayer.setAudioStreamType(AudioManager.STREAM_ALARM)
        }
        mediaPlayer.setDataSource(this.context, uri)
        mediaPlayer.prepare()
        val f = volume / VOLUME_SCALE
        mediaPlayer.setVolume(f, f)
        return mediaPlayer
    }

    private fun reuseExistingPlayer(
        player: MediaPlayer,
        uri: Uri,
        volume: Int,
    ): MediaPlayer {
        Log.d(TAG, "Reusing existing Audio Player")
        player.reset()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            player.setAudioAttributes(
                AudioAttributes
                    .Builder()
                    .setUsage(AudioAttributes.USAGE_ALARM)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build(),
            )
        } else {
            @Suppress("DEPRECATION")
            player.setAudioStreamType(AudioManager.STREAM_ALARM)
        }
        player.setDataSource(this.context, uri)
        player.prepare()
        val f = volume / VOLUME_SCALE
        player.setVolume(f, f)
        return player
    }

    private fun stopAndRelease() {
        Log.d(TAG, "stopAndRelease Audio Instance")
        this.player?.let { p ->
            try {
                p.stop()
                p.reset()
            } catch (e: IllegalStateException) {
                Log.d(TAG, e.message ?: "")
            }
            try {
                p.release()
            } catch (e: IllegalStateException) {
                Log.d(TAG, e.message ?: "")
            }
            this.player = null
        }
    }

    suspend fun playAbsVolume(
        uri: Uri?,
        volume: Int,
    ) {
        if (uri == null) {
            Log.e(TAG, "Cannot play null URI")
            return
        }
        withContext(Dispatchers.IO) {
            val p = this@Audio.player
            this@Audio.player =
                if (p != null) {
                    reuseExistingPlayer(p, uri, volume)
                } else {
                    preparePlayer(uri, volume)
                }
        }
        val p = this.player ?: return
        this.playing = true
        Log.d(TAG, "Start playing Bell")
        p.isLooping = false
        p.start()
        p.setOnCompletionListener(this)
    }

    fun isPlaying(): Boolean = this.playing

    fun getStreamVolume(): Int {
        val streamMaxVolume = this.manager.getStreamMaxVolume(AudioManager.STREAM_ALARM)
        val streamVolume = this.manager.getStreamVolume(AudioManager.STREAM_ALARM).toFloat()
        val round = Math.round((VOLUME_SCALE * streamVolume) / streamMaxVolume)
        Log.d(TAG, "getStreamVolume: vol=$streamVolume max=$streamMaxVolume = $round")
        return round
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        Log.d(TAG, "onCompletion called")
        this.playing = false
        stopAndRelease()
    }

    companion object {
        private const val TAG = "ZMT_Audio"

        private const val VOLUME_SCALE = 100.0f
    }
}
